Fix: DO-2905: Migrate all components to Floating UI, fix rerenders, remove popper dependency

Roman-Kornev opened the pull request:
<!--- The title format is expected to follow the pattern:-->
<!--- CHANGE TYPE: Ticket Name -->
<!--- Docs: DO-100 Update PR Template -->
<!--- Where CHANGE TYPEs are: Docs, Breaking, Improvement, Fix, Refactor, Feat -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it has a spec, please link to the spec here. -->
Continuing #98, this PR migrates the rest of the components, fixing rerendering issues in the process.
Full list of components now using Floating UI instead of Popper:
- `MultiSelect`
- `Select`
- `ComboBox`
- `ContextMenu`
- `DatepickerSelect`
- `SectionedList`
- `Table.OptionsMenu`

All components were rerendering all options on every scroll event, even when closed. The only exception was Select, which only started to rerender when opened, but then still continued to rerender all items even after being closed. For a large list this caused huge performance issues.
Here is a benchmark of just scrolling a page with a closed Select component with 1000 elements:
BEFORE:
![image](https://github.com/causalens/dara-ui/assets/119343173/ff35db6a-90fc-438c-acd0-76929fffa5c2)

AFTER:
![image](https://github.com/causalens/dara-ui/assets/119343173/c96027b1-b3e6-477e-9a3d-7c17ea4db891)


Notice the `Scripting` time on the chart for a 1000x difference. On average, the total render time for just 1 such Select was cut down from 26ms to ~5ms, so an 80% reduction in render time.

## Implementation Description
<!--- Why did you follow this implementation approach- -->
<!--- Is there any background knowledge necessary to understand the approach taken- -->
<!--- Describe the limitations, pitfalls and tradeoffs of the approach- -->
- New middleware `matchWidthToReference`, similar to the old `sameWidthModifier`, matches the width of the floating element to be the same as the width of the reference element. Optionally accepts a delta to the width.
- `ContextMenu` offset changed from (-2, +4) to (+4, +4) for consistency, should also prevent accidental action triggers on double-right-clicks.
- `DatepickerSelect`'s width is still hardcoded at `16.25rem`, as it doesn't depend on the `Select`'s width but the whole `Datepicker`'s width. A more proper fix would be to pass a header ref and use clientWidth, but our styling of ReactDatepicker shrinks it further, so the width won't be exact. Decided to not touch the structure and leave it as is.
- The previous implementation was implicitly relying on the rerender of `Table.OptionsMenu` to clear the selected values. It now uses a `key` prop to force a reset of state. The alternative is to store the selected item state in `Table.OptionsMenu`, but it's not used anywhere else, so a `key` prop sounds more appropriate here.
- All components now use `useRole(context, { role: <ROLE> })` for better accessibility.

## Any new dependencies Introduced
<!--- Where there any dependencies added? Why?- -->
Removed `@popperjs/core` and `react-popper` dependencies.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Locally with a modified storybook to allow for scrolling (check one of the latest reverts for examples)

## PR Checklist:
<!--- Go over all the following points, and ensure it has all been checked with an `x` -->

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [x] I have added relevant tests (unit, integration or regression).
- [x] I have added comments to all the bits that are hard to follow.
- [x] I have added/updated Documentation.
- [x] I have updated the appropriate changelog with a line for my changes.

## Screenshots (if appropriate):
<!--- For UI changes make sure to add an image or GIF showcasing the change-->

patricia-causalens commented:
Hi Roman, great work, this is an awesome improvement!

Was playing with it locally and noticed that now the dropdown items don't always match the width of the select:
<img width="708" alt="Screenshot 2024-04-30 at 09 35 46" src="https://github.com/causalens/dara-ui/assets/104359539/7ddffba2-d656-4a39-8e07-8d2d4fc8739b">
<img width="708" alt="Screenshot 2024-04-30 at 09 36 28" src="https://github.com/causalens/dara-ui/assets/104359539/86b58856-4725-4249-a931-25da2a240350">

patricia-causalens reviewed (changes requested):
Hi Roman, great work, this is an awesome improvement!

Was playing with it locally and noticed that now the dropdown items don't always match the width of the select:
<img width="708" alt="Screenshot 2024-04-30 at 09 35 46" src="https://github.com/causalens/dara-ui/assets/104359539/604fc3bb-dada-4302-a470-22f8501a78f0">
<img width="708" alt="Screenshot 2024-04-30 at 09 36 28" src="https://github.com/causalens/dara-ui/assets/104359539/5c98a831-3921-43b8-9b5d-5f9af944d0e4">

krzysztof-causalens commented on packages/ui-components/src/sectioned-list/sectioned-list.tsx:
```diff
@@ -310,13 +309,13 @@ function SectionedList(props: SectionedListProps): JSX.Element {
-                    {...attributes.popper}
+                    {...getFloatingProps()}
+                    ref={mergedRefs}
+                    role="listbox"
```
Doesn't `getFloatingProps` include a role?

Roman-Kornev replied:
That's true, but that was causing tests to fail. It turns out that it's because downshift already includes those aria labels on the inner component, so we ended up with 2 nested aria comboboxes. 

I moved the reference props to the inner element and supplied a proper `useRole(context, { role: <ROLE> })`. That effectively merges those props which solves this issue, but I'm not sure if having the `ref` and its `{...getReferenceProps()}` attached to different elements is that great either.

krzysztof-causalens reviewed (approved):
Looks good, I have a question regarding the roles

Roman-Kornev commented:
@patricia-causalens, it looks like the modifier was not applied to selects as `applySameWidthModifier` was checking specifically for `=== false` before, even though the comment says it's true by default. Changed it to actually be `true` by default.

Also added an optional delta-width to the modifier and removed the width parsing logic entirely. All components should now match the width in the default configuration.

patricia-causalens reviewed (approved):
Very nice, LGTM!
